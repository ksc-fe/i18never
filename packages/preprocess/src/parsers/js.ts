import { parse as babelParse, ParseResult } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import {
    ImportDeclaration,
    ObjectProperty,
    StringLiteral,
    TemplateLiteral,
    Context as BaseContext,
    Text,
    SourceLocation,
} from '@i18never/shared';
import * as t from '@babel/types';

type TextNode = t.StringLiteral | t.TemplateLiteral | t.JSXText;
export type Entity = NodePath<TextNode>;
export type Context = BaseContext<TextNode>;

// const defaultRootLoc: SourceLocation = { line: 1, column: 0 };
export function parse(
    source: string,
    // because the column of location of babel parser starts from 0, but the others, e.g.
    // Vue/Pug, start from 1. We use a defaultRootLoc starting from 1 to keep
    // consistent with them.
    rootLoc?: SourceLocation
) {
    const ast = babelParse(source, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
    });

    return getContext(ast, rootLoc).keys;
}

function getContext(ast: ParseResult<t.File>, rootLoc?: SourceLocation) {
    const context: Context = {
        keys: [],
        rootLoc,
    };

    traverse(
        ast,
        {
            ImportDeclaration,
            ObjectProperty,
            StringLiteral,
            TemplateLiteral,
            JSXText: Text<t.JSXText>,
        },
        undefined,
        context
    );

    return context;
}

// function formatJsLocLine(jsKey: TempKeyItem, rootLine: number) {
//     // - 2 is script index ande template index
//     const line = jsKey.loc.line + rootLine - 2;
//     const column = jsKey.jsx ? jsKey.loc.column : jsKey.loc.column + 1;

//     return {
//         line,
//         column,
//     };
// }

// function getAllKeysLoc(ast, filename: string) {
//     const allKeys: TempKeyItem[] = [];

//     traverse(ast, visitor, undefined, {
//         keys: allKeys,
//         filename,
//     });

//     // const code = generate(ast, {
//     //     // concise: true,
//     //     jsescOption: { minimal: true },
//     //     retainLines: true,
//     // }).code;
//     // console.log('log', code);
//     return allKeys;
// }
