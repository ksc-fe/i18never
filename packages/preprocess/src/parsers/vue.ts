import {
    baseParse as vueBaseParse,
    NodeTypes,
    Node as VueNode,
    InterpolationNode,
    SimpleExpressionNode,
    TextNode,
    ElementNode,
    AttributeNode,
    DirectiveNode,
} from '@vue/compiler-core';
import { parse as jsParse } from './js';
import { KeyItem, pugParse } from './';
import { parseString, SourceLocation, getLoc } from '@i18never/shared';
import { parse as vueParse } from '@vue/compiler-sfc';

export type Entity = TextNode;
export type VueVisitors = {
    SimpleExpression(node: SimpleExpressionNode): void;
    Text(node: TextNode): void;
};

export function parse(source: string) {
    const keys: KeyItem[] = [];
    const { descriptor } = vueParse(source);
    const { template, script, scriptSetup } = descriptor;

    if (template) {
        const rootLoc = getLoc(template.loc.start, null, 1);
        if (template.lang === 'pug') {
            keys.push(...pugParse(template.content, rootLoc));
        } else {
            keys.push(...getTemplateKeys(template.ast));
        }
    }

    if (script) {
        keys.push(
            ...jsParse(script.content, getLoc(script.loc.start, null, 1))
        );
    }

    if (scriptSetup) {
        keys.push(
            ...jsParse(
                scriptSetup.content,
                getLoc(scriptSetup.loc.start, null, 1)
            )
        );
    }

    return keys;
}

export function parseTemplate(value: string, rootLoc?: SourceLocation) {
    return getTemplateKeys(vueBaseParse(value), rootLoc);
}

export function parseSimpleExpression(
    value: string,
    loc: SourceLocation,
    rootLoc?: SourceLocation
) {
    // we must use a pair of bracket to wrap the value, so that the js parser
    // can parse it. The startNumber should also add 1 (1 + 1 = 2)
    // i.e.: { a: true } => ({ a: true })
    return jsParse(`(${value})`, getLoc(loc, rootLoc, 2));
}

function getTemplateKeys(ast: VueNode | undefined, rootLoc?: SourceLocation) {
    const keys: KeyItem[] = [];
    walkTemplateNode(ast, {
        SimpleExpression(node) {
            keys.push(
                ...parseSimpleExpression(node.content, node.loc.start, rootLoc)
            );
        },
        Text(node) {
            const content = node.content;
            if (!content.trim()) return;

            const nodeLoc = node.loc;
            const loc = getLoc(nodeLoc.start, rootLoc, 1);
            if (nodeLoc.source !== content) {
                // if source isn't equal to content, it indicates
                // the content has quotes, so forward one column
                // i.e. <div a="测试"></div>
                loc.column++;
            }

            keys.push({
                ...parseString(content),
                loc,
                entity: node,
            });
        },
    });

    return keys;
}

function walkTemplateNode(node: VueNode | undefined, visitors: VueVisitors) {
    if (!node) return;

    switch (node.type) {
        case NodeTypes.ROOT:
        case NodeTypes.ELEMENT:
            if (node.type === NodeTypes.ELEMENT) {
                (node as ElementNode).props.forEach((attr) => {
                    walkTemplateNode(attr, visitors);
                });
            }
            (node as ElementNode).children.forEach((node) =>
                walkTemplateNode(node, visitors)
            );
            break;
        case NodeTypes.TEXT:
            visitors.Text(node as TextNode);
            break;
        case NodeTypes.INTERPOLATION:
            walkTemplateNode((node as InterpolationNode).content, visitors);
            break;
        case NodeTypes.SIMPLE_EXPRESSION:
            visitors.SimpleExpression(node as SimpleExpressionNode);
            break;
        case NodeTypes.ATTRIBUTE:
            walkTemplateNode((node as AttributeNode).value, visitors);
            break;
        case NodeTypes.DIRECTIVE:
            walkTemplateNode((node as DirectiveNode).exp, visitors);
            break;
    }
}
