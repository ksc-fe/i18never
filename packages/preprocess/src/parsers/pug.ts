import pugParse from 'pug-parser';
import { Lexer } from 'pug-lexer';
import pugWalk from 'pug-walk';
import { parseString, SourceLocation, getLoc } from '@i18never/shared';
import { KeyItem } from './';
import { parseTemplate } from './vue';
import { parse as jsParse } from './js';

type BaseNode = { line: number; column: number };
type TextNode = BaseNode & { val: string; type: 'Text' };
type TagNode = BaseNode & { attrs: AttrNode[]; type: 'Tag' };
type AttrNode = BaseNode & { name: string; val: string };
type Node = TextNode | TagNode;
export type Entity = TextNode | AttrNode;

export function parse(source: string, rootLoc?: SourceLocation) {
    const lexer = new Lexer(source);
    const tokens = lexer.getTokens();
    const ast = pugParse(tokens);
    const keys: KeyItem[] = [];

    pugWalk(ast, function before(node: Node) {
        switch (node.type) {
            case 'Text':
                keys.push(...parseTemplate(node.val, getLoc(node, rootLoc, 1)));
                break;
            case 'Tag':
                node.attrs.forEach((node: AttrNode) => {
                    let value = node.val;
                    if (!value) return;

                    // should remove quotes
                    value = value.slice(1, -1);

                    const name = node.name;
                    const firstChar = name[0];
                    const loc = {
                        line: node.line,
                        column: node.column + name.length + 2, // `${node.name}="`.length,
                    };
                    if (firstChar === ':' || firstChar === '@') {
                        keys.push(...jsParse(value, loc));
                    } else {
                        keys.push({
                            ...parseString(value),
                            loc: getLoc(loc, rootLoc, 1),
                            entity: node,
                        });
                    }
                });
                break;
        }
    });

    return keys;
}
