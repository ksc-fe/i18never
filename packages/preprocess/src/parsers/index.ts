import { KeyItem as BaseKeyItem } from '@i18never/shared';
import { Entity as JsEntity } from './js';
import { Entity as PugEntity } from './pug';
import { Entity as VueEntity } from './vue';
export { parse as jsParse } from './js';
export { parse as pugParse } from './pug';
export { parse as vueParse } from './vue';

export type KeyItem = BaseKeyItem<JsEntity | PugEntity | VueEntity>;
