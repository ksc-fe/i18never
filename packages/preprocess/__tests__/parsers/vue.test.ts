import { parse } from '../../src/parsers/vue';
import { getToMatchSnapshot } from './toMatchSnapshot';

const toMatchSnapshot = getToMatchSnapshot(parse);

test('static template', () => {
    toMatchSnapshot(`<template><div>测试</div></template>`);
});

test('interpolation', () => {
    toMatchSnapshot(`<template><div>{{ '测试' + a }}</div></template>`);
});

test('template with props', () => {
    toMatchSnapshot(`<template><div a="测试"></div></template>`);
});

test('template with expression props', () => {
    toMatchSnapshot(`<template><div :a="'测试' + a"></div></template>`);
});

test('pug template', () => {
    toMatchSnapshot(`<template lang="pug">div 测试</template>`);
});

test('script', () => {
    toMatchSnapshot(`<script>const a = "测试"</script>`);
});

test('script setup', () => {
    toMatchSnapshot(`<script setup>const a = "测试"</script>`);
});

test(`template with newline`, () => {
    toMatchSnapshot(`<template>
        <div>测试</div>
    </template>`);
});

test(`multiple lines with props`, () => {
    toMatchSnapshot(`<template>
    <div>
        <Form a="测试"></Form>
    </div>
</template>`);
});

test('prop with object value', () => {
    toMatchSnapshot(
        `<template><div :a="{a: '测试', b: true}"></div></template>`
    );
});

test('Multi line coordinate actual value verification', () => {
    let keys =  parse(
        `<script setup>
            const a = test
        </script>
        <template>
            <div :a="{a: '测试', b: true}"></div>
        </template>`
    );
    expect(keys[0].loc).toStrictEqual({
        'line': 5,
        'column': 26
    })
});
