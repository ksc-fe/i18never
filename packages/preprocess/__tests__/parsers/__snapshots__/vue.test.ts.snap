// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpolation 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 29,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`multiple lines with props 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 17,
      "line": 3,
    },
    "tags": null,
  },
]
`;

exports[`pug template 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 25,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`script 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 19,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`script setup 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 25,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`static template 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 25,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`template with expression props 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 30,
      "line": 1,
    },
    "tags": null,
  },
]
`;

exports[`template with newline 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 13,
      "line": 2,
    },
    "tags": null,
  },
]
`;

exports[`template with props 1`] = `
Array [
  Object {
    "identifier": null,
    "key": "测试",
    "loc": Object {
      "column": 28,
      "line": 1,
    },
    "tags": null,
  },
]
`;
